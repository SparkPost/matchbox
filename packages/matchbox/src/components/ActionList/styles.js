import styled from 'styled-components';
import { tokens } from '@sparkpost/design-tokens';
import { UnstyledLink } from '../UnstyledLink';
import { Box } from '../Box';
import { buttonReset } from '../../styles/helpers';

export const StyledSection = styled(Box)`
  border-top: ${props => props.theme.borders['300']};
  padding-bottom: ${tokens.spacing_200};
  padding-top: ${tokens.spacing_200};

  &:first-child {
    border: none;
    padding-top: ${tokens.spacing_0};
  }

  &:last-child {
    padding-bottom: ${tokens.spacing_0};
  }
`;

export const StyledLink = styled(UnstyledLink)`
  ${props => (props.isType ? buttonReset : '')}
  display: block;
  width: 100%;
  padding: ${tokens.spacing_300} ${tokens.spacing_400};
  transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
  transition-property: color, background;
  cursor: pointer;
  font-size: ${tokens.fontSize_300};
  line-height: ${tokens.lineHeight_300};
  text-align: left;

  &,
  &:visited {
    text-decoration: none;
    color: ${props => (props.highlighted ? tokens.color_blue_700 : tokens.color_gray_900)};
    background: ${props => (props.highlighted ? tokens.color_blue_200 : 'none')};
  }

  &:hover,
  &:focus {
    outline: none;
    color: ${tokens.color_gray_900};
    background: ${tokens.color_blue_100};
  }

  ${'' /* This only applies to buttons, not links */}
  &:disabled {
    opacity: 0.6;
    pointer-events: none;
  }
`;
