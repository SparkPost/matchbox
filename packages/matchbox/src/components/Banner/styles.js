import { tokens } from '@sparkpost/design-tokens';
import { ErrorOutline, CheckCircleOutline, InfoOutline } from '@sparkpost/matchbox-icons';

export function container(props) {
  switch (props.status) {
    case 'success':
      return `
        background: ${tokens.color_green_100};
        border: 1px solid ${tokens.color_green_600};
      `;
    case 'warning':
      return `
        background: ${tokens.color_yellow_100};
        border: 1px solid ${tokens.color_yellow_500};
      `;
    case 'danger':
      return `
        background: ${tokens.color_red_100};
        border: 1px solid ${tokens.color_red_700};
      `;
    // Gray banner no longer exists as of 4.0.0
    case 'default':
    default:
      return `
        background: ${tokens.color_blue_100};
        border: 1px solid ${tokens.color_blue_600};
      `;
  }
}

// Overwrites unstyled link colors within Banner content to be color-contrast accessible
export function childLinks() {
  return `
  p, ul, ol {
    a, a:visited {
      color: ${tokens.color_blue_800};
      &:hover {
        color: ${tokens.color_blue_900};
      }
    }
  }
  `;
}

export const statusIcons = {
  success: {
    iconName: CheckCircleOutline,
    iconLabel: 'Success',
    bg: 'green.600',
    fill: 'white',
  },
  info: {
    iconName: InfoOutline,
    iconLabel: 'Info',
    bg: 'blue.700',
    fill: 'white',
  },
  warning: {
    iconName: ErrorOutline,
    iconLabel: 'Warning',
    bg: 'yellow.300',
    fill: 'yellow.700',
  },
  danger: {
    iconName: ErrorOutline,
    iconLabel: 'Error',
    bg: 'red.700',
    fill: 'white',
  },
};

export function dismissBase() {
  return `
    padding: 0.25rem;
    transition: background ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
    &:hover {
      cursor: pointer;
    }
  `;
}

export function dismissColor(props) {
  switch (props.status) {
    case 'success':
      return `
        &:hover {
          background: ${tokens.color_green_300};
        }
      `;
    case 'warning':
      return `
        &:hover {
          background: ${tokens.color_yellow_300};
        }
      `;
    case 'danger':
      return `
        &:hover {
          background: ${tokens.color_red_300};
        }
      `;
    // Gray banner no longer exists as of 4.0.0
    case 'default':
    default:
      return `
        &:hover {
          background: ${tokens.color_blue_300};
        }
      `;
  }
}
