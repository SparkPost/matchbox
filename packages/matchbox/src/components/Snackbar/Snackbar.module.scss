@import '../../styles/config';

.Snackbar {
  position: relative;
  display: inline-block;

  border-radius: border-radius(large);
  box-shadow: shadow(deep);
  background: color(gray, 2);

  color: color(gray, 10);
}

.success {
  background: color(green);
}

.danger, .error {
  background: color(red);
}

.warning {
  background: color(yellow, light);
  color: color(gray, 1);

  .Content {
    a, a:visited {
      color: color(gray, 1);
    }
  }

  .Dismiss {
    color: color(gray, 1);
  }

  .DismissIcon {
    fill: color(gray, 1);
  }
}

.Content {
  display: inline-block;
  vertical-align: top;
  margin-right: -4px;
  padding: rem(12) spacing();
  min-width: rem(200);
  font-weight: 500;
  font-size: font-size(500);

  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  a, a:visited {
    color: color(gray, 10);
    text-decoration: underline;
  }
}

.Dismiss {
  display: inline-block;
  margin-right: -4px;
  vertical-align: top;
  padding: rem(12) spacing();
  border-bottom: none;
  user-select: none;

  color: color(gray, 10);
  font-size: font-size(200);

  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  &:hover {
    .DismissIcon {
      opacity: 1;
    }
  }
}

.DismissIcon {
  fill: color(gray, 10);
  opacity: 0.9;
  transition: opacity 0.15s;
}
