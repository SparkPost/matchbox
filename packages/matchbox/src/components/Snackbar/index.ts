export { default as Snackbar } from './Snackbar';
