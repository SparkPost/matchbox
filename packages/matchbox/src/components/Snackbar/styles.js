import { tokens } from '@sparkpost/design-tokens';

export const base = () => `
  display: inline-flex;
  align-items: flex-start;
  color: ${tokens.color_white};
`;

export const status = props => {
  switch (props.status) {
    case 'success':
      return `background: ${tokens.color_green_800};`;
    case 'danger':
    case 'error':
      return `background: ${tokens.color_red_700};`;
    case 'warning':
      return `
        background: ${tokens.color_yellow_300};
        color: ${tokens.color_gray_900};
      `;
    case 'default':
    default:
      return `background: ${tokens.color_blue_800};`;
  }
};

export const dismiss = () => `
  display: inline-flex;
  border: none;
  background: transparent;
  transition: background ${tokens.motionDuration_fast};
`;

export const dismissStatus = props => {
  let color;

  switch (props.status) {
    case 'success':
      color = tokens.color_green_900;
      break;
    case 'danger':
    case 'error':
      color = tokens.color_red_800;
      break;
    case 'warning':
      color = tokens.color_yellow_400;
      break;
    case 'default':
    default:
      color = tokens.color_blue_900;
      break;
  }

  return `
    &:hover {
      background: ${color};
    }
  `;
};
