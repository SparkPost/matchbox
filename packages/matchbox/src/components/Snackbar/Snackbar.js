import React from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import PropTypes from 'prop-types';
import { createPropTypes } from '@styled-system/prop-types';
import { Close, Info, CheckCircle, Warning, ErrorIcon } from '@sparkpost/matchbox-icons';
import { margin, maxWidth as maxWidthSystem } from 'styled-system';
import { base, status, dismiss, dismissStatus } from './styles';
import { buttonReset } from '../../styles/helpers';
import { pick } from '../../helpers/props';

const StyledBox = styled(Box)`
  ${base}
  ${status}
  ${margin}
`;

const StyledClose = styled(Box)`
  ${buttonReset}
  ${dismiss}
  ${dismissStatus}
`;

const Snackbar = React.forwardRef(function Snackbar(props, ref) {
  const { children, status, maxWidth, onDismiss, ...rest } = props;
  const systemProps = pick(rest, margin.propNames);

  return (
    <StyledBox
      status={status}
      role="alert"
      p="300"
      borderRadius="100"
      ref={ref}
      tabIndex="-1"
      {...systemProps}
    >
      <Box mr="300">
        {status === 'default' && <Info size="20" label="Info" />}
        {status === 'success' && <CheckCircle size="20" label="Success" />}
        {status === 'warning' && <Warning size="20" label="Warning" />}
        {(status === 'error' || status === 'danger') && <ErrorIcon size="20" label="Error" />}
      </Box>
      <Box fontSize="200" lineHeight="200" py="100" pr="400" maxWidth={maxWidth} minWidth="12.5rem">
        {children}
      </Box>
      <StyledClose
        status={status}
        as="button"
        p="0.125rem"
        mx="100"
        onClick={onDismiss}
        type="button"
      >
        <Close size={24} />
        <ScreenReaderOnly>Close</ScreenReaderOnly>
      </StyledClose>
    </StyledBox>
  );
});

Snackbar.displayName = 'Snackbar';
Snackbar.defaultProps = {
  status: 'default',
  maxWidth: 380,
};
Snackbar.propTypes = {
  /**
   * The type of snackbar.
   */
  status: PropTypes.oneOf(['default', 'success', 'danger', 'error', 'warning']),

  /**
   * Callback when dismiss button is clicked.
   */
  onDismiss: PropTypes.func.isRequired,

  /**
   * Snackbar Content
   */
  children: PropTypes.node,
  ...createPropTypes(margin.propNames),
  ...createPropTypes(maxWidthSystem.propNames),
};

export default Snackbar;
