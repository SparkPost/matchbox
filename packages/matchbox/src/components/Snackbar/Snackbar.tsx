import React from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { Close, Info, CheckCircle, Warning, ErrorIcon } from '@sparkpost/matchbox-icons';
import { margin, MarginProps, MaxWidthProps } from 'styled-system';
import css from '@styled-system/css';
import { base, status, dismiss, dismissStatus } from './styles';
import { buttonReset } from '../../styles/helpers';
import { pick } from '../../helpers/props';

type StyledButtonProps = {
  $status?: 'default' | 'success' | 'danger' | 'warning' | 'error';
};

const StyledBox = styled(Box)<StyledButtonProps>`
  ${base}
  ${status}
  ${margin}
`;

const StyledClose = styled.button<StyledButtonProps>`
  ${buttonReset}
  ${css({
    p: '0.125rem',
    mx: '100',
  })}
  ${dismiss}
  ${dismissStatus}
`;

export type SnackbarProps = {
  'data-id'?: string;
  status?: 'default' | 'success' | 'danger' | 'warning' | 'error';
  onDismiss?: React.MouseEventHandler<HTMLButtonElement>;
  children?: React.ReactNode;
} & MarginProps &
  MaxWidthProps;

const Snackbar = React.forwardRef<HTMLDivElement, SnackbarProps>(function Snackbar(props, userRef) {
  const {
    children,
    status = 'default',
    maxWidth = 380,
    onDismiss,
    'data-id': dataId,
    ...rest
  } = props;
  const systemProps = pick(rest, margin.propNames);

  return (
    <StyledBox
      $status={status}
      role="alert"
      p="300"
      borderRadius="100"
      ref={userRef}
      tabIndex={-1}
      {...systemProps}
      data-id={dataId}
    >
      <Box mr="300">
        {status === 'default' && <Info size="20" label="Info" />}
        {status === 'success' && <CheckCircle size="20" label="Success" />}
        {status === 'warning' && <Warning size="20" label="Warning" />}
        {(status === 'error' || status === 'danger') && <ErrorIcon size="20" label="Error" />}
      </Box>
      <Box fontSize="200" lineHeight="200" py="100" pr="400" maxWidth={maxWidth} minWidth="12.5rem">
        {children}
      </Box>
      <StyledClose $status={status} onClick={onDismiss} type="button">
        <Close size={24} />
        <ScreenReaderOnly>Close</ScreenReaderOnly>
      </StyledClose>
    </StyledBox>
  );
}) as React.ForwardRefExoticComponent<SnackbarProps>;

Snackbar.displayName = 'Snackbar';

export default Snackbar;
