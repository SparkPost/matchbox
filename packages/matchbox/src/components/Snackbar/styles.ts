import { tokens } from '@sparkpost/design-tokens';
import React from 'react';
import Snackbar from './Snackbar';

export const base = () => `
  display: inline-flex;
  align-items: flex-start;
  color: ${tokens.color_white};
`;

export const status = ({ status }: Pick<React.ComponentProps<typeof Snackbar>, 'status'>) => {
  function makeLinkColorStyles(color) {
    return `
      a, 
      a:hover, 
      a:visited { 
        color: ${color} !important; 
      }
    `;
  }
  const whiteLinks = makeLinkColorStyles(tokens.color_white);
  const greyLinks = makeLinkColorStyles(tokens.color_gray_900);

  switch (status) {
    case 'success':
      return `
        ${whiteLinks} 
        background: ${tokens.color_green_800};
      `;
    case 'danger':
    case 'error':
      return `
        ${whiteLinks} background: ${tokens.color_red_700};
      `;
    case 'warning':
      return `
        ${greyLinks} 
        background: ${tokens.color_yellow_300};
        color: ${tokens.color_gray_900};
      `;
    case 'default':
    default:
      return `
        ${whiteLinks} background: ${tokens.color_blue_800};
      `;
  }
};

export const dismiss = () => `
  display: inline-flex;
  transition: background ${tokens.motionDuration_fast};
  &:hover {
    cursor: pointer;
  }
`;

export const dismissStatus = ({
  status,
}: Pick<React.ComponentProps<typeof Snackbar>, 'status'>) => {
  let color;

  switch (status) {
    case 'success':
      color = tokens.color_green_900;
      break;
    case 'danger':
    case 'error':
      color = tokens.color_red_800;
      break;
    case 'warning':
      color = tokens.color_yellow_400;
      break;
    case 'default':
    default:
      color = tokens.color_blue_900;
      break;
  }

  return `
    &:hover {
      background: ${color};
    }
  `;
};
