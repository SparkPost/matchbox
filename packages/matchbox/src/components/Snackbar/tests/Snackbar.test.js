import React from 'react';
import { shallow } from 'enzyme';

import Snackbar from '../Snackbar';
import styles from '../Snackbar.module.scss';

describe('Snackbar', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      status: 'default',
      maxWidth: 500,
      onDismiss: jest.fn()
    };
    wrapper = shallow(<Snackbar {...props}>Snacksssss</Snackbar>);
  });

  it('renders Snackbar', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Snackbar with different props', () => {
    wrapper.setProps({ maxWidth: 100, status: 'danger' });
    expect(wrapper).toMatchSnapshot();
  });

  it('invokes onDismiss', () => {
    wrapper.find(`.${styles.Dismiss}`).simulate('click');
    expect(props.onDismiss).toHaveBeenCalledTimes(1);
  });

});
