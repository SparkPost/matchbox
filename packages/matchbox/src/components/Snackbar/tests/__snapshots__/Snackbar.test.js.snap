// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar renders Snackbar 1`] = `
<div
  className="Snackbar default"
  role="alert"
>
  <div
    className="Content"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    Snacksssss
  </div>
  <a
    className="Dismiss"
    href="javascript:void(0);"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
  >
    <Close
      className="DismissIcon"
      size={21}
    />
  </a>
</div>
`;

exports[`Snackbar renders Snackbar with different props 1`] = `
<div
  className="Snackbar danger"
  role="alert"
>
  <div
    className="Content"
    style={
      Object {
        "maxWidth": 100,
      }
    }
  >
    Snacksssss
  </div>
  <a
    className="Dismiss"
    href="javascript:void(0);"
    onClick={[MockFunction]}
    onKeyDown={[Function]}
    role="button"
  >
    <Close
      className="DismissIcon"
      size={21}
    />
  </a>
</div>
`;
